src/compiler/crystal/semantic/call.cr

```diff
@@ -361,7 +361,19 @@ class Crystal::Call
     end
   end
 
-  def instantiate(matches, owner, self_type, with_literals)
+  def instantiate(signature, matches, owner, self_type, with_literals)
+    if with_literals
+      # Now that we have all our matches, check if any of them matches exactly
+      # all types, assuming autocasted values will always match (because they
+      # matches and they were not ambiguous). If so, only keep matches up to
+      # that exact match. We need to do this here because with autocasting
+      # we consider all overloads to detect ambiguous usage.
+      stop_index = matches.index do |match|
+        signature.matches_exactly?(match, with_literals: true)
+      end
+      matches = matches[..stop_index] if stop_index
```

Comment:
If there is an exact match why the matches array does not contain a single element?

Reply:
When autocasting is not considered, because there can't be an ambiguity, as soon as a match is an exact match there's no need to see other methods.

But with autocasting we need to check if there's ambiguity. For example:

```crystal
def fill(x : Int64, y : Int)
end

def fill(x : Int8, y : Int)
end

fill(0, 0)
```

That will match the first method, but also the second, and that's when you get the "ambiguous call" error.

If there are no ambiguities, it means all methods matched... but maybe one of them was an exact match? We need to do this after we considered all methods to check for ambiguity.

Ideally, `lookup_matches` would do this, but then we need to pass `with_literals` to those methods, and they are many. It's simpler to do it here.

Did this answer the question?

I think autocasting is a nice feature but it makes the method lookup very complex. Maybe we can remove it from 1.0?